// react-interface/es/components/FlexTable/Cell.js
var _templateObject = _taggedTemplateLiteralLoose(['\n  display: flex;\n  flex-direction: ', ';\n  flex-wrap: ', ';\n  flex-grow: ', ';\n  flex-shrink: ', ';\n  flex-basis: ', ';\n  justify-content: ', ';\n  align-items: ', ';\n  padding: 0.5em;\n  word-break: break-word;\n  overflow: hidden;\n  text-overflow: ellipsis;\n  min-width: 0px;\n  ', '\n  ', '\n'], ['\n  display: flex;\n  flex-direction: ', ';\n  flex-wrap: ', ';\n  flex-grow: ', ';\n  flex-shrink: ', ';\n  flex-basis: ', ';\n  justify-content: ', ';\n  align-items: ', ';\n  padding: 0.5em;\n  word-break: break-word;\n  overflow: hidden;\n  text-overflow: ellipsis;\n  min-width: 0px;\n  ', '\n  ', '\n']);

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _taggedTemplateLiteralLoose(strings, raw) { strings.raw = raw; return strings; }

import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { space, width } from 'styled-system';

var Wrapper = styled.div(_templateObject, function (props) {
  return props.direction;
}, function (props) {
  return props.wrap;
}, function (props) {
  return props.grow;
}, function (props) {
  return props.shrink;
}, function (props) {
  return props.basis;
}, function (props) {
  return props.justify;
}, function (props) {
  return props.align;
}, space, width);

var Cell = function Cell(_ref) {
  var children = _ref.children,
      props = _objectWithoutProperties(_ref, ['children']);

  return React.createElement(
    Wrapper,
    props,
    children
  );
};

Cell.defaultProps = {
  grow: 1,
  shrink: 1,
  basis: 'auto',
  align: 'initial',
  justify: 'start',
  wrap: 'nowrap',
  direction: 'row'
};

Cell.propTypes = process.env.NODE_ENV !== "production" ? {
  grow: PropTypes.number,
  basis: PropTypes.any,
  align: PropTypes.string,
  children: PropTypes.node,
  justify: PropTypes.string,
  wrap: PropTypes.string,
  direction: PropTypes.string
} : {};

export default Cell;
