// react-interface/es/components/Select/Wrapper.js
var _templateObject = _taggedTemplateLiteralLoose(['\n  display: flex;\n  align-items: center;\n  justify-content: center;\n  height: 100%;\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n\n  svg {\n    color: ', ';\n  }\n\n  &:focus {\n    border: 1px solid ', ';\n    outline: none;\n  }\n'], ['\n  display: flex;\n  align-items: center;\n  justify-content: center;\n  height: 100%;\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n\n  svg {\n    color: ', ';\n  }\n\n  &:focus {\n    border: 1px solid ', ';\n    outline: none;\n  }\n']);

function _taggedTemplateLiteralLoose(strings, raw) { strings.raw = raw; return strings; }

import styled from 'styled-components';
import { space, width, borderRadius, borders, borderColor, color, hover } from 'styled-system';

var Wrapper = styled.div(_templateObject, space, width, borderRadius, borders, borderColor, color, hover, function (props) {
  return props.theme.colors.selectIcon;
}, function (props) {
  return props.theme.colors.selectFocusBorder;
});

Wrapper.defaultProps = {
  p: 2,
  color: 'select',
  bg: 'selectBackground',
  borderColor: 'selectBorder',
  borderRadius: 2,
  border: '1px solid',
  hover: {
    color: 'selectHover',
    backgroundColor: 'selectHoverBackground'
  }
};

export default Wrapper;
