// react-interface/es/components/Popper/Popper.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _templateObject = _taggedTemplateLiteralLoose(['\n  .popper {\n    z-index: 99;\n    overflow-y: scroll;\n    max-height: ', ';\n    display: ', ';\n    border-width: ', ';\n    border-style: ', ';\n    border-color: ', ';\n    box-shadow: ', ';\n    ', '\n    ', '\n    ', '\n    ', '\n    ', '\n    ', '\n    ', '\n  }\n\n  .popper .popper__arrow {\n    width: 0;\n    height: 0;\n    border-style: solid;\n    position: absolute;\n    margin: 5px;\n  }\n\n  .popper[data-placement^="top"].has-arrow {\n    margin-bottom: 5px;\n  }\n\n  .popper[data-placement^="top"] .popper__arrow {\n    border-width: 5px 5px 0 5px;\n    border-color: #FFF transparent transparent transparent;\n    box-shadow: rgba(0, 0, 0, 0.2) 1px 1px 1px;\n    bottom: -5px;\n    // left: calc(50% - 5px);\n    margin-top: 0;\n    margin-bottom: 0;\n  }\n\n  .popper[data-placement^="bottom"].has-arrow {\n    margin-top: 5px;\n  }\n\n  .popper[data-placement^="bottom"] .popper__arrow {\n    border-width: 0 5px 5px 5px;\n    border-color: transparent transparent #FFF transparent;\n    box-shadow: rgba(0, 0, 0, 0.2) 1px 1px 1px;\n    top: -5px;\n    // left: calc(50% - 5px);\n    margin-top: 0;\n    margin-bottom: 0;\n  }\n\n  .popper[data-placement^="right"].has-arrow {\n    margin-left: 5px;\n  }\n\n  .popper[data-placement^="right"] .popper__arrow {\n    border-width: 5px 5px 5px 0;\n    border-color: transparent #FFF transparent transparent;\n    box-shadow: rgba(0, 0, 0, 0.2) 1px 1px 1px;\n    left: -5px;\n    top: calc(50% - 5px);\n    margin-left: 0;\n    margin-right: 0;\n  }\n\n  .popper[data-placement^="left"].has-arrow {\n    margin-right: 5px;\n  }\n\n  .popper[data-placement^="left"] .popper__arrow {\n    border-width: 5px 0 5px 5px;\n    border-color: transparent transparent transparent #FFF;\n    box-shadow: rgba(0, 0, 0, 0.2) 1px 1px 1px;\n    right: -5px;\n    top: calc(50% - 5px);\n    margin-left: 0;\n    margin-right: 0;\n  }\n\n  // .popper[data-x-out-of-boundaries] {\n  //   display: none;\n  // }\n'], ['\n  .popper {\n    z-index: 99;\n    overflow-y: scroll;\n    max-height: ', ';\n    display: ', ';\n    border-width: ', ';\n    border-style: ', ';\n    border-color: ', ';\n    box-shadow: ', ';\n    ', '\n    ', '\n    ', '\n    ', '\n    ', '\n    ', '\n    ', '\n  }\n\n  .popper .popper__arrow {\n    width: 0;\n    height: 0;\n    border-style: solid;\n    position: absolute;\n    margin: 5px;\n  }\n\n  .popper[data-placement^="top"].has-arrow {\n    margin-bottom: 5px;\n  }\n\n  .popper[data-placement^="top"] .popper__arrow {\n    border-width: 5px 5px 0 5px;\n    border-color: #FFF transparent transparent transparent;\n    box-shadow: rgba(0, 0, 0, 0.2) 1px 1px 1px;\n    bottom: -5px;\n    // left: calc(50% - 5px);\n    margin-top: 0;\n    margin-bottom: 0;\n  }\n\n  .popper[data-placement^="bottom"].has-arrow {\n    margin-top: 5px;\n  }\n\n  .popper[data-placement^="bottom"] .popper__arrow {\n    border-width: 0 5px 5px 5px;\n    border-color: transparent transparent #FFF transparent;\n    box-shadow: rgba(0, 0, 0, 0.2) 1px 1px 1px;\n    top: -5px;\n    // left: calc(50% - 5px);\n    margin-top: 0;\n    margin-bottom: 0;\n  }\n\n  .popper[data-placement^="right"].has-arrow {\n    margin-left: 5px;\n  }\n\n  .popper[data-placement^="right"] .popper__arrow {\n    border-width: 5px 5px 5px 0;\n    border-color: transparent #FFF transparent transparent;\n    box-shadow: rgba(0, 0, 0, 0.2) 1px 1px 1px;\n    left: -5px;\n    top: calc(50% - 5px);\n    margin-left: 0;\n    margin-right: 0;\n  }\n\n  .popper[data-placement^="left"].has-arrow {\n    margin-right: 5px;\n  }\n\n  .popper[data-placement^="left"] .popper__arrow {\n    border-width: 5px 0 5px 5px;\n    border-color: transparent transparent transparent #FFF;\n    box-shadow: rgba(0, 0, 0, 0.2) 1px 1px 1px;\n    right: -5px;\n    top: calc(50% - 5px);\n    margin-left: 0;\n    margin-right: 0;\n  }\n\n  // .popper[data-x-out-of-boundaries] {\n  //   display: none;\n  // }\n']);

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _taggedTemplateLiteralLoose(strings, raw) { strings.raw = raw; return strings; }

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { findDOMNode } from 'react-dom';
import styled from 'styled-components';
import { borders, color, borderColor, boxShadow, borderRadius, space, width } from 'styled-system';
import resizeDetector from 'element-resize-detector';
import cx from 'classnames';
import { Manager, Target, Popper as Positioner, Arrow } from 'react-popper';
// import Show from 'react-show'
import { debounce } from '../../utils/browser';
import Portal from '../Portal';

var PopoverWrapper = styled.div(_templateObject, function (props) {
  return props.maxHeight;
}, function (props) {
  return props.isOpen ? 'block' : 'none';
}, function (props) {
  return props.theme['popover.border.width'];
}, function (props) {
  return props.theme['popover.border.style'];
}, function (props) {
  return props.theme['popover.border.color'];
}, function (props) {
  return props.theme['popover.shadow'];
}, color, borders, borderColor, borderRadius, boxShadow, space, width);

var animations = {
  scale: {
    default: { opacity: 0 },
    hide: {
      opacity: 0,
      transform: 'scale(.8) translateY(-30%)',
      pointerEvents: 'none'
    },
    show: { opacity: 1, transform: 'none', pointerEvents: 'auto' }
  },
  fade: {
    default: { opacity: 0 },
    hide: { opacity: 0 },
    show: { opacity: 1 }
  },
  slide: {
    default: { opacity: 0 },
    hide: { opacity: 0, transform: 'translateY(-2em)' },
    show: { opacity: 1, transform: 'translateY(0)' }
  }
};

var CustomTarget = function CustomTarget(_ref) {
  var innerRef = _ref.innerRef,
      props = _objectWithoutProperties(_ref, ['innerRef']);

  return React.createElement('div', _extends({ ref: innerRef, style: { cursor: 'pointer' } }, props));
};

var CustomPopper = function CustomPopper(_ref2) {
  var innerRef = _ref2.innerRef,
      props = _objectWithoutProperties(_ref2, ['innerRef']);

  return React.createElement('div', _extends({ ref: innerRef }, props, { style: _extends({}, props.style) })) // left: 1 if needed
  ;
};

var Popper = function (_PureComponent) {
  _inherits(Popper, _PureComponent);

  function Popper(props) {
    _classCallCheck(this, Popper);

    var _this = _possibleConstructorReturn(this, _PureComponent.call(this, props));

    _this._setOutsideTap = function () {
      document.addEventListener('click', _this._handleOutsideTap, true);
    };

    _this._setResizeDetector = function () {
      _this.detector = resizeDetector();

      if (_this.target) {
        _this.detector.listenTo(_this.target, _this._handleResize);
      }
    };

    _this._handleResize = function (el) {
      _this.mostRecentWidth = _this.state.width;
      _this.setState({
        width: el.offsetWidth
      });
    };

    _this._handleOutsideTap = function (e) {
      var clickedPopper = e && _this.popper && _this.popper.contains(e.target);
      var clickedTarget = e && _this.target && _this.target.contains(e.target);

      if (!e || !clickedPopper && !clickedTarget) {
        _this.close();
      }
    };

    _this.close = function () {
      _this.setState({ isOpen: false }, _this.props.onClose);
    };

    _this._handleTargetClick = function () {
      if (_this.state.isOpen) {
        _this.close();
      } else {
        _this.setState({ isOpen: true }, _this.props.onOpen);
      }
    };

    _this._handleResize = debounce(_this._handleResize, 50, false);

    _this.state = {
      isOpen: false,
      width: null
    };
    return _this;
  }

  Popper.prototype.componentDidUpdate = function componentDidUpdate(lastProps, lastState) {
    var _this2 = this;

    if (!lastState.isOpen && this.state.isOpen) {
      setTimeout(function () {
        return _this2._setOutsideTap();
      });
      setTimeout(function () {
        return _this2._setResizeDetector();
      });
    }
  };

  Popper.prototype.componentWillUnmount = function componentWillUnmount() {
    document.removeEventListener('click', this._handleOutsideTap, true);
    if (this.detector) {
      this.detector.uninstall();
    }
  };

  Popper.prototype._getPopperWidth = function _getPopperWidth() {
    if (this.target && this.props.fullWidth) {
      if (!this.state.width) {
        return this.target.getBoundingClientRect().width - 2;
      } else {
        return this.state.width - 2;
      }
    }

    if (this.props.width) {
      return this.props.width;
    }
  };

  Popper.prototype.renderPositioner = function renderPositioner() {
    var _this3 = this;

    var _props = this.props,
        position = _props.position,
        easing = _props.easing,
        duration = _props.duration,
        animation = _props.animation,
        children = _props.children,
        modifiers = _props.modifiers,
        arrow = _props.arrow,
        contentStyle = _props.contentStyle;


    var popperClasses = cx({
      popper: true,
      'has-arrow': arrow
    });

    var popperStyle = _extends({}, contentStyle, {
      width: this._getPopperWidth(),
      zIndex: 99,
      display: this.state.isOpen ? 'block' : 'none'
    });

    var popperModifiers = _extends({
      preventOverflow: {
        enabled: true,
        escapeWithReference: true,
        boundariesElement: 'scrollParent'
      }
    }, modifiers);

    return React.createElement(
      Positioner,
      {
        key: 'popper',
        placement: position,
        modifiers: popperModifiers,
        className: popperClasses,
        offset: '-100px',
        component: CustomPopper,
        innerRef: function innerRef(c) {
          _this3.popper = findDOMNode(c);
        }
      },
      React.createElement(
        'div',
        { style: popperStyle },
        typeof children === 'function' ? children(this.close, this.state.isOpen) : children,
        arrow && React.createElement(Arrow, { className: 'popper__arrow' })
      )
    );
  };

  Popper.prototype.render = function render() {
    var _this4 = this;

    var _props2 = this.props,
        trigger = _props2.trigger,
        portal = _props2.portal,
        rest = _objectWithoutProperties(_props2, ['trigger', 'portal']);

    return React.createElement(
      PopoverWrapper,
      _extends({}, rest, { isOpen: this.state.isOpen }),
      React.createElement(
        Manager,
        null,
        React.createElement(
          Target,
          {
            innerRef: function innerRef(c) {
              return _this4.target = findDOMNode(c);
            },
            component: CustomTarget,
            onClick: this._handleTargetClick
          },
          trigger
        ),
        portal ? React.createElement(
          Portal,
          null,
          this.renderPositioner()
        ) : this.renderPositioner()
      )
    );
  };

  return Popper;
}(PureComponent);

// <Show
//   show={this.state.isOpen}
//   easing={easing}
//   duration={duration}
//   style={animations[animation].default}
//   styleHide={animations[animation].hide}
//   styleShow={animations[animation].show}
//   transitionOnMount
// >
//   <div style={popperStyle}>
//     {typeof children === 'function'
//       ? children(this._handleOutsideTap, this.state.isOpen)
//       : children}
//     {arrow && <Arrow className="popper__arrow" />}
//   </div>
// </Show>

Popper.propTypes = process.env.NODE_ENV !== "production" ? {
  arrow: PropTypes.bool,
  animation: PropTypes.oneOf(['fade', 'scale', 'slide']),
  duration: PropTypes.number,
  offset: PropTypes.number,
  onOpen: PropTypes.func,
  onClose: PropTypes.func,
  fullWidth: PropTypes.bool,
  portal: PropTypes.bool,
  contentStyle: PropTypes.object,
  trigger: PropTypes.oneOfType([PropTypes.func, PropTypes.element]).isRequired,
  on: PropTypes.oneOfType([PropTypes.oneOf(['hover', 'click', 'focus']), PropTypes.arrayOf(PropTypes.oneOf(['hover', 'click', 'focus']))]),
  children: PropTypes.oneOfType([PropTypes.func, PropTypes.element, PropTypes.string]).isRequired,
  position: PropTypes.oneOf(['top', 'top-start', 'top-end', 'bottom', 'bottom-start', 'bottom-end', 'right', 'right-start', 'right-end', 'left', 'left-start', 'left-end', 'auto', 'auto-start', 'auto-end'])
} : {};

Popper.defaultProps = {
  onOpen: function onOpen() {},
  onClose: function onClose() {},
  closeOnDocumentClick: true,
  defaultOpen: false,
  on: ['click'],
  portal: false,
  arrow: true,
  animation: 'slide',
  easing: 'easeOutQuint',
  duration: 250,
  position: 'bottom-start',
  fullWidth: false,
  contentStyle: {}
};

export default Popper;
