// react-interface/es/components/Spinner/Spinner.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _templateObject = _taggedTemplateLiteralLoose(['\n  from {transform: rotate(0deg)}\n  to {transform: rotate(360deg)}\n'], ['\n  from {transform: rotate(0deg)}\n  to {transform: rotate(360deg)}\n']),
    _templateObject2 = _taggedTemplateLiteralLoose(['\n  ', ' svg {\n    ', ' transition-property: transform;\n    animation-name: ', ';\n    animation-iteration-count: infinite;\n    animation-timing-function: linear;\n    color: ', ';\n    ', ';\n  }\n'], ['\n  ', ' svg {\n    ', ' transition-property: transform;\n    animation-name: ', ';\n    animation-iteration-count: infinite;\n    animation-timing-function: linear;\n    color: ', ';\n    ', ';\n  }\n']);

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _taggedTemplateLiteralLoose(strings, raw) { strings.raw = raw; return strings; }

// Could make a timber one: https://codepen.io/zsherman/pen/JMQzjJ for inspiration
import React from 'react';
import PropTypes from 'prop-types';
import styled, { keyframes } from 'styled-components';
import { space, width, color } from 'styled-system';

var spin = keyframes(_templateObject);

function speedSwitch(speed) {
  if (speed === 'fast') return 600;
  if (speed === 'slow') return 900;
  return 750;
}

var Wrapper = styled.div(_templateObject2, space, width, spin, function (props) {
  return props.theme.colors.spinner;
}, color);

var Spinner = function Spinner(_ref) {
  var color = _ref.color,
      speed = _ref.speed,
      gap = _ref.gap,
      thickness = _ref.thickness,
      size = _ref.size,
      style = _ref.style,
      rest = _objectWithoutProperties(_ref, ['color', 'speed', 'gap', 'thickness', 'size', 'style']);

  return React.createElement(
    Wrapper,
    _extends({ color: color, speed: speed, style: style, fontSize: size }, rest),
    React.createElement(
      'svg',
      {
        style: { animationDuration: speedSwitch(speed) + 'ms' },
        role: 'img',
        'aria-labelledby': 'title desc',
        viewBox: '0 0 32 32',
        stroke: 'currentColor'
      },
      React.createElement('circle', {
        role: 'presentation',
        cx: 16,
        cy: 16,
        r: 14 - thickness / 2,
        fill: 'none',
        strokeWidth: thickness,
        strokeDasharray: Math.PI * 2 * (11 - gap),
        strokeLinecap: 'round'
      })
    )
  );
};

Spinner.propTypes = process.env.NODE_ENV !== "production" ? {
  color: PropTypes.string,
  thickness: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8]).isRequired,
  gap: PropTypes.oneOf([1, 2, 3, 4, 5]).isRequired,
  speed: PropTypes.oneOf(['fast', 'slow']),
  height: PropTypes.number,
  width: PropTypes.number
} : {};

Spinner.defaultProps = {
  color: 'spinner',
  gap: 4,
  thickness: 4,
  height: 20,
  width: 20,
  speed: 'fast'
};

export default Spinner;
