// react-interface/es/components/MultiSelect/Wrapper.js
var _templateObject = _taggedTemplateLiteralLoose(['\n  background: ', ';\n  border: 1px solid ', ';\n  display: flex;\n  align-items: center;\n  justify-content: flex-start;\n  flex-wrap: wrap;\n  height: 100%;\n  padding: 10px;\n  position: relative;\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n\n  &:focus {\n    border: 1px solid ', ';\n    outline: none;\n  }\n\n  ', '\n  ', '\n\n  svg {\n    color: ', ';\n  }\n'], ['\n  background: ', ';\n  border: 1px solid ', ';\n  display: flex;\n  align-items: center;\n  justify-content: flex-start;\n  flex-wrap: wrap;\n  height: 100%;\n  padding: 10px;\n  position: relative;\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n\n  &:focus {\n    border: 1px solid ', ';\n    outline: none;\n  }\n\n  ', '\n  ', '\n\n  svg {\n    color: ', ';\n  }\n']);

function _taggedTemplateLiteralLoose(strings, raw) { strings.raw = raw; return strings; }

import styled from 'styled-components';
import { space, width, borderRadius, color, borders, borderColor, hover, focus } from 'styled-system';

var Wrapper = styled.div(_templateObject, function (props) {
  return props.theme.colors.multiSelectBackground;
}, function (props) {
  return props.theme.colors.multiSelectBorder;
}, space, width, borderRadius, color, borders, borderColor, function (props) {
  return props.theme.colors.multiSelectFocusBorder;
}, focus, hover, function (props) {
  return props.theme.colors.multiSelectIcon;
});

Wrapper.defaultProps = {
  p: 2,
  color: 'multiSelect',
  bg: 'multiSelectBackground',
  borderColor: 'multiSelectBorder',
  borderRadius: 2,
  border: '1px solid',
  hover: {
    color: 'multiSelectHover',
    backgroundColor: 'multiSelectHoverBackground'
  }
};

export default Wrapper;
