// react-interface/es/components/Button/Button.js
var _templateObject = _taggedTemplateLiteralLoose(['\n  font-smoothing: antialiased;\n  display: inline-block;\n  vertical-align: middle;\n  text-align: center;\n  text-decoration: none;\n  font-family: inherit;\n  cursor: pointer;\n  outline: none;\n  transition: all .2s ease-in-out;\n  display: inline-flex;\n  align-items: center;\n  justify-content: center;\n\n  /* normalize borders */\n  border-width: 1px;\n  border-style: solid;\n  border-radius: 3px;\n\n  ', '\n\n  ', '\n\n  ', '\n\n  ', '\n\n  ', '\n\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n\n  &:disabled {\n    opacity: .25;\n    background: ', ';\n    color: ', ';\n  }\n'], ['\n  font-smoothing: antialiased;\n  display: inline-block;\n  vertical-align: middle;\n  text-align: center;\n  text-decoration: none;\n  font-family: inherit;\n  cursor: pointer;\n  outline: none;\n  transition: all .2s ease-in-out;\n  display: inline-flex;\n  align-items: center;\n  justify-content: center;\n\n  /* normalize borders */\n  border-width: 1px;\n  border-style: solid;\n  border-radius: 3px;\n\n  ', '\n\n  ', '\n\n  ', '\n\n  ', '\n\n  ', '\n\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n\n  &:disabled {\n    opacity: .25;\n    background: ', ';\n    color: ', ';\n  }\n']),
    _templateObject2 = _taggedTemplateLiteralLoose(['\n      background: ', ';\n      color: ', ';\n      border-color: ', ';\n    '], ['\n      background: ', ';\n      color: ', ';\n      border-color: ', ';\n    ']),
    _templateObject3 = _taggedTemplateLiteralLoose(['\n      background: none;\n      color: ', ';\n      border: none;\n    '], ['\n      background: none;\n      color: ', ';\n      border: none;\n    ']),
    _templateObject4 = _taggedTemplateLiteralLoose(['\n      font-size: ', ';\n      padding: ', ';\n    '], ['\n      font-size: ', ';\n      padding: ', ';\n    ']);

function _taggedTemplateLiteralLoose(strings, raw) { strings.raw = raw; return strings; }

import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import { darken, lighten } from 'polished';
import { space, width, color, borders, borderRadius, borderColor, fontWeight, fontSize, hover, boxShadow } from 'styled-system';

var Button = styled.button(_templateObject, function (props) {
  return !props.palette && css(_templateObject2, function (props) {
    return props.theme.colors.buttonBackground;
  }, function (props) {
    return props.theme.colors.button;
  }, function (props) {
    return props.theme.colors.buttonBorder;
  });
}, function (props) {
  return props.palette && css(_templateObject2, function (props) {
    return props.theme.colors[props.palette][3];
  }, function (props) {
    return props.theme.colors[props.palette][0];
  }, function (props) {
    return props.theme.colors[props.palette][4];
  });
}, function (props) {
  return props.inverse && props.palette && css(_templateObject2, function (props) {
    return props.theme.colors[props.palette][0];
  }, function (props) {
    return props.theme.colors[props.palette][4];
  }, function (props) {
    return props.theme.colors[props.palette][1];
  });
}, function (props) {
  return props.transparent && css(_templateObject3, function (props) {
    return props.palette ? props.theme.colors[props.palette][4] : 'inherit';
  });
}, function (props) {
  return props.size && css(_templateObject4, function (props) {
    return props.theme.fontSizes[props.theme.sizes[props.size]];
  }, function (props) {
    return props.theme.buttonPaddings[props.theme.sizes[props.size]];
  });
}, color, space, width, borders, borderRadius, borderColor, fontWeight, fontSize, hover, boxShadow, function (props) {
  return props.theme.colors.buttonDisabledBackground;
}, function (props) {
  return props.theme.colors.buttonDisabled;
});

Button.propTypes = {
  /**
   * What type of component is this?
   */
  palette: PropTypes.oneOf(['primary', 'gray', 'red', 'orange', 'yellow', 'green', 'teal', 'blue', 'indigo', 'purple', 'pink', 'info', 'danger', 'warning', 'success', PropTypes.string]),
  /**
   * How big is it?
   */
  size: PropTypes.oneOf(['xxs', 'xs', 'sm', 'md', 'lg', 'xl', 'xxl', 'xxxl']),
  /**
   * Style overrides
   */
  styles: PropTypes.shape({
    color: PropTypes.string,
    border: PropTypes.string
  }),
  /**
   * Inverse colors
   */
  inverse: PropTypes.bool,
  /**
   * Transparent
   */
  transparent: PropTypes.bool
};

Button.defaultProps = {
  size: 'md',
  inverse: false,
  transparent: false

  /** @component */
};export default Button;
