// react-interface/es/components/Notification/Notification.js
var _templateObject = _taggedTemplateLiteralLoose(['\n  display: flex;\n  align-items: center;\n  justify-content: start;\n  line-height: 1rem;\n  border-width: 1px;\n  border-style: solid;\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n\n  ', '\n\n  ', '\n'], ['\n  display: flex;\n  align-items: center;\n  justify-content: start;\n  line-height: 1rem;\n  border-width: 1px;\n  border-style: solid;\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n  ', '\n\n  ', '\n\n  ', '\n']),
    _templateObject2 = _taggedTemplateLiteralLoose(['\n      background: ', ';\n      color: ', ';\n      border-color: ', ';\n    '], ['\n      background: ', ';\n      color: ', ';\n      border-color: ', ';\n    ']);

function _taggedTemplateLiteralLoose(strings, raw) { strings.raw = raw; return strings; }

import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import { space, borderRadius, color, justifyContent, width, fontSize, borders } from 'styled-system';

var Notification = styled.div(_templateObject, borderRadius, borders, color, space, justifyContent, width, fontSize, function (props) {
  return props.palette && css(_templateObject2, function (props) {
    return props.theme.colors[props.palette][0];
  }, function (props) {
    return props.theme.colors[props.palette][4];
  }, function (props) {
    return props.theme.colors[props.palette][2];
  });
}, function (props) {
  return props.palette && props.solid && css(_templateObject2, function (props) {
    return props.theme.colors[props.palette][3];
  }, function (props) {
    return props.theme.colors[props.palette][0];
  }, function (props) {
    return props.theme.colors[props.palette][4];
  });
});

Notification.defaultProps = {
  borderRadius: 4,
  p: 3,
  solid: false,
  bg: 'notificationBackground',
  color: 'notificationColor',
  borderColor: 'notificationBorder'

  /** @component */
};export default Notification;
